File: iosApp/iosApp/Auth/OtpPassVerify/OtpPassVerify.swift
//
//  OtpPassVerify.swift
//  Experiment
//
//  Created by Abhishek Rathore on 11/12/24.
//

import SwiftUI
import Shared

struct OtpPassVerify: View {
    @EnvironmentObject private var navigation: Navigation
    @Environment(\.dismiss) private var dismiss
    @ObservedObject private var viewModel = ViewModelProvider.shared.otpPasswordViewModel
    
    init(email: String){
        viewModel.doInitEmail(email: email)
    }
    
    var body: some View{
        CircularIndicatorBox(isLoading: viewModel.state.isLoading, content: {
            OtpView(title: "Password Reset",
                    email: viewModel.getEmail(),
                    error: viewModel.state.error,
                    onOtpChange: {value in viewModel.onOtpChange(otp: value)},
                    resendAction: {viewModel.resendOtp()},
                    verifyAction: {viewModel.verifyPassReset()}
            )
        })
        .navigationBarBackButtonHidden()
        .toolbar {
            ToolbarItem(placement: .navigationBarLeading){
                BackButton{
                    dismiss()
                    navigation.popBackStack()
                }
            }
        }
        .onDisappear{
            viewModel.resetState()
        }
        .onChange(of: viewModel.state.navigateScreen) { _, newValue in
            if newValue {
                navigation.navigateTo(
                    destination: NavRoutes.ResetPassword(email: viewModel.getEmail()),
                    popUpToBuilder: {
                        navigation.navigateToStartDestination()
                    }
                )
                viewModel.resetNavigate()
            }
        }
    }
}

extension OtpPassVerifyViewModel{
    var state: OtpUi{
        get{
            return self.state(
                \.otpPassVerifyUiState,
                 equals: { $0 == $1 },
                 mapper: { $0 }
            )
        }
    }
    
    func resetState(){
        resetNavigate()
    }
}

#Preview {
    OtpPassVerify(email: "Example@gmail.com")
        .environmentObject(Navigation())
}

File: composeApp/src/androidMain/kotlin/org/example/project/kmmchat/presentation/auth/OtpPassVerify.kt
package org.example.project.kmmchat.presentation.auth

import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import org.koin.androidx.compose.koinViewModel

@Composable
fun OtpPassVerify(
    email: String,
    navigateResetPassword: () -> Unit,
    navigateBack: () -> Unit,
    viewModel: OtpPassVerifyViewModel = koinViewModel()
) {
    LaunchedEffect(Unit) {
        viewModel.initEmail(email = email)
    }
    val otpUiState by viewModel.otpPassVerifyUiState.collectAsStateWithLifecycle()
    LaunchedEffect(key1 = otpUiState) {
        if(otpUiState.navigateScreen){
            navigateResetPassword()
            viewModel.resetNavigate()
        }
    }
    Otp(
        title = "OTP For Password Reset",
        email = email,
        otpUiState = otpUiState,
        navigateBack = navigateBack,
        onChangeOtp = { otp -> viewModel.onOtpChange(otp) },
        onSubmit = { viewModel.verifyPassReset() },
        resendOtp = { viewModel.resendOtp() }
    )
}